### DISCUSSION TITLE
Should Amethyst be allowed to send application names to custom layouts?

### DISCUSSION THREAD

**Discussion by ianyh** [ianyh/Amethyst#1425] [Q&A] [answered]
As it stands now we explicitly do _not_ pass application information along with window frames when asking custom layout for appropriate frames. A user was requesting something for custom layouts in which they could apply special layout rules to some lower priority applications that should not affect the tiling. With the app information restriction this is effectively impossible, but I wanted to ask for feedback on how acceptable it would be to include application names along with window frames sent to custom layouts.

A brief explanation for _why_ that information is not included is that Amethyst has privileged access to information through Accessibility, and custom layouts are basically arbitrary JavaScript so I wanted to minimize how much privileged information was passed along. Any application can determine what other applications are running, but they cannot necessarily tie those applications to window frames.

There is information that I don't want to ever pass along, but the open question right now is whether or not tying window frames to application names is acceptable. What do people think?

**Comment by lominming** [comment]
Trying to understand the main scenario that is causing the concern. Is the concern that user will accidentally download and run a malicious custom javascript layout file? Amethyst already self-selects more advanced users. On top of that, custom layouts are even more advanced.

I am definitely more in favor of more information of the windows because that is how we can setup advanced tiling rules. I think of tiling manager like automation, and automation in general meant that more information = more degree of automation. 

A possibility is a checkbox just like the one available for debugging where user requires explicit step to allow passing of more details.

**Comment by ianyh** [comment]
> Trying to understand the main scenario that is causing the concern. Is the concern that user will accidentally download and run a malicious custom javascript layout file? Amethyst already self-selects more advanced users. On top of that, custom layouts are even more advanced.
> 
> 
> 
> I am definitely more in favor of more information of the windows because that is how we can setup advanced tiling rules. I think of tiling manager like automation, and automation in general meant that more information = more degree of automation. 
> 
> 
> 
> A possibility is a checkbox just like the one available for debugging where user requires explicit step to allow passing of more details. 

The issue isn't so much accidental, but rather that anything can add a custom layout and modify Amethyst preferences to use the layout. They are both essentially just user-writable files. A checkbox would be backed by the preferences file so it has the same problem. You have to assume that anything that gets passed to a custom layout is potentially remotely available.

I might just be overly paranoid. Tying application names to window frames is probably harmless. I don't want to do window titles, though, as that actually does open up some side channels.

**Comment by thavidu** [comment]
I would really really like this feature too! I'm trying to make a custom layout where certain windows/apps are just permanently docked in certain positions and everything else is left free-floating (or using one of the tiled layouts)

**Reply by denisoby** [reply]
I would like to have the same feature with permanent locked layouts as @thavidu mentioned.

For example
- layout "Web development"
screen 1
... chrome taking 60% space
... dev Tools taking 40%

screen 2
... vscode 100%

screen 3
... console1 taking 50%
... console2 taking 50%

- layout "Backend development"
less chrome
+1 vscode
+1 wireshark


- layout "Unit Tests"
- layout "Zoom call"
...

I'm regular manually have to launch/reposition windows and would really appreciate any way to automatically switch these environments.

**Comment by ianyh** [comment, marked as answer]
Conclusion from the poll is that it's desired.

**Reply by HelloThisIsFlo** [reply]
+1

Application names would allow me to recreate the "snapshot" feature of Moom that I use quite a lot. That is a layout with a predefined arrangement of windows in predefined positions. I used one for my morning daily review, it's very useful. Window titles would be even more useful, but I understand if you're weary about it.
